activerecord/lib/active_record/scoping/default.rb

```diff
@@ -107,7 +107,11 @@ def build_default_scope(base_rel = nil)
 
             if default_scope_override
               # The user has defined their own default scope method, so call that
-              evaluate_default_scope { default_scope }
+              evaluate_default_scope do
+                if scope = default_scope
+                  (base_rel ||= relation).merge(scope)
```

Comment:
Yes, it has failed. The previous version of this PR overwrites `current_scope` when `build_default_scope` is already affected in scoping.

```
% ARCONN=sqlite3 be ruby -w -Itest test/cases/inheritance_test.rb
Using sqlite3
Run options: --seed 45762

# Running:

........F.......................................................

Finished in 0.576700s, 110.9762 runs/s, 235.8245 assertions/s.

  1) Failure:
InheritanceTest#test_inheritance_with_default_scope_respects_scoping [test/cases/inheritance_test.rb:446]:
Expected: 0
  Actual: 1

64 runs, 136 assertions, 1 failures, 0 errors, 0 skips
```

```diff
diff --git a/activerecord/lib/active_record/scoping/default.rb b/activerecord/lib/active_record/scoping/default.rb
index 70b2693b28..7f14daac70 100644
--- a/activerecord/lib/active_record/scoping/default.rb
+++ b/activerecord/lib/active_record/scoping/default.rb
@@ -108,9 +108,7 @@ def build_default_scope(base_rel = nil)
             if default_scope_override
               # The user has defined their own default scope method, so call that
               evaluate_default_scope do
-                if scope = default_scope
-                  (base_rel ||= relation).merge(scope)
-                end
+                default_scope
               end
             elsif default_scopes.any?
               base_rel ||= relation
diff --git a/activerecord/lib/active_record/scoping/named.rb b/activerecord/lib/active_record/scoping/named.rb
index a61fdd6454..7fb198d913 100644
--- a/activerecord/lib/active_record/scoping/named.rb
+++ b/activerecord/lib/active_record/scoping/named.rb
@@ -31,7 +31,7 @@ def all
 
         def default_scoped # :nodoc:
           scope = relation
-          build_default_scope(scope) || scope
+          scope.scoping { build_default_scope(scope) || scope }
         end
 
         def default_extensions # :nodoc:
```